lib/molinillo/resolution.rb

```diff
@@ -481,16 +441,41 @@ def push_state_for_requirements(new_requirements, requires_sort = true, new_acti
       # @return [Array] possibilities
       def possibilities_for_requirement(requirement, activated = self.activated)
         return [] unless requirement
-        locked_requirement = locked_requirement_named(name_for(requirement))
+        if locked_requirement_named(name_for(requirement))
+          return locked_requirement_possibility_set(requirement, activated)
+        end
+
+        # Build an array of PossibilitySets, with each element representing a group of
+        # dependency versions that all have the same sub-dependency version constraints.
+        possibility_sets = []
+        possibility_sets_index = {}
+
+        search_for(requirement).reverse_each do |possibility|
+          dependencies = dependencies_for(possibility)
+          if index = possibility_sets_index[dependencies]
+            possibility_sets[index].possibilities.unshift(possibility)
+          else
+            possibility_sets << PossibilitySet.new(dependencies, [possibility])
+            possibility_sets_index[dependencies] = possibility_sets.count - 1
+          end
+        end
+
+        possibility_sets.reverse
+      end
+
+      # @param [Object] the proposed requirement
+      # @return [Array] possibility set containing only the locked requirement, if any
+      def locked_requirement_possibility_set(requirement, activated = self.activated)
         all_possibilities = search_for(requirement)
-        return all_possibilities unless locked_requirement
+        locked_requirement = locked_requirement_named(name_for(requirement))
 
         # Longwinded way to build a possibilities array with either the locked
         # requirement or nothing in it. Required, since the API for
         # locked_requirement isn't guaranteed.
-        all_possibilities.select do |possibility|
+        locked_possibility = all_possibilities.find do |possibility|
```

Comment:
I think this needs to be a `select` -- theres no guarantee that the search only yields a single value for a locked spec